README.md
# Carrot-Catcher

A short game I made in about a week.
I was introduced to the engine, Love2D, one day, and a week later this project emerged just to prove I could do something and finish it with this new engine on my own.

To play, download everything except the src folder. Your firewall may give you some problems, but I promise it isn'ta virus.
The source files are included in the src folder. You can compile it yourself if you so wish. Information on that can be found on the website for Love2D.

This is also the first game made in the NumberAnimalz universe of games.


changelog.txt
v1.0.1 (Nov. 10th, 12018)
-Due to a problem where frames spontaneously drop, the player could go out of bounds, which then causes you to be unable to move. The out of bounds part was fixed, but the random FPS drops still happen. Not sure why.
-Previously, the carrots would fall and be seen as underneath the rhino. This is no longer the case. The rhino does not float above the carrots.
-The version was changed from 1.0.0 to 1.0.1 after 1 day of release.
-Sounds are now stored in memory rather than streamed. I thought this would fix the FPS drop at least a tiny bit, but it didn't.

v1.0.2 (Mar. 3rd, 12019)
-Love2D updated from 11.1 to 11.2, so we updated the dependencies.
-Some users were experiencing issues where the program wasn't starting. In an effort to fix this, love.exe has also been provided in the zip, as well as a bat file that runs Carrot Catcher directly through love.exe.
-The version was change from 1.0.1 to 1.0.2 after 120 days of release.

src/bats.lua
bats={}
batsLoaded=false

function bats.load()
	bats.sprite=love.graphics.newImage("images/bats.png")
	batWidth=bats.sprite:getHeight()
	batHeight=bats.sprite:getWidth()
	
	bats.radiusFromCenter = 550
	bats.angleInRadians=((2*math.pi*(math.random(1,360)))/360)
	bats.xPos=(bats.radiusFromCenter*math.cos(bats.angleInRadians) + 310)
	bats.yPos=(bats.radiusFromCenter*math.sin(bats.angleInRadians) + 233)
	bats.xCompR=0
	bats.yCompR=0
	bats.velocity=200
	bats.xVel=0
	bats.yVel=0
	bats.satisfied=false
	
	bats.xCompR=(player.x-bats.xPos)
	bats.yCompR=(player.y-bats.yPos)
	bats.xyAngle=math.atan2(bats.xCompR,bats.yCompR)
	bats.xVel=(bats.velocity*math.sin(bats.xyAngle))
	bats.yVel=(bats.velocity*math.cos(bats.xyAngle))
end

function bats.update(dt)
	if batsLoaded==false then
		bats.load()
		batsLoaded=true
	end
	bats.xPos=(bats.xPos+bats.xVel*dt)
	bats.yPos=(bats.yPos+bats.yVel*dt)
	
	
	------------
	if CheckCollision(bats.xPos,bats.yPos,batWidth,batHeight,player.x,player.y,player.sprite:getWidth(),player.sprite:getHeight()) and pause == false and bats.satisfied==false then
		pause=true
		bats.satisfied=true
		player.carrotsCollected=player.carrotsCollected-25
		collisionSound:play()
		if player.carrotsCollected < 0 then
			gameOver=true
			player.carrotsCollected = 0
		end
		
		pause=false
	elseif ((((bats.xPos-310)^2 + (bats.yPos-233)^2)^(1/2))>555) and pause == false then
		pause=true
		
		batsLoaded=false
		BatsReleased=false
		pause = false
	end
	---------------
end

function bats.draw()
	love.graphics.draw(bats.sprite,bats.xPos,bats.yPos)
end

src/rats.lua
rats={}
rats.list={}
ratSpeed=200
absMaxRats=4
maxRats=1
fourHundredReached=false
fiveHundredReached=false

function rats.load()
	ratSprite=love.graphics.newImage("images/rat.png")
	rightRatSprite=love.graphics.newImage("images/rat2.png")
	ratWidth=ratSprite:getWidth()
	ratHeight=ratSprite:getHeight()
	
	for i=1,maxRats do
		rats.list[i] = {}	
		rats.list[i].x=(-200-ratWidth*1.5)
		rats.list[i].speed=ratSpeed
		rats.list[i].y=(math.random(0,570-ratHeight))
		rats.list[i].direction="left"
	end
end

function rats.update(dt)
	for i,v in pairs (rats.list) do
		if (rats.list[i].direction=="left") then
			rats.list[i].x = (rats.list[i].speed*dt + rats.list[i].x)
		else 
			rats.list[i].x = (rats.list[i].x - rats.list[i].speed*dt)
		end
	end
	
	for i,v in pairs (rats.list) do
		if CheckCollision(rats.list[i].x,rats.list[i].y,ratWidth,ratHeight,player.x,player.y,player.sprite:getWidth(),player.sprite:getHeight()) and pause == false then
			pause=true
			player.carrotsCollected=player.carrotsCollected-10*beingTrampled
			collisionSound:play()
			if player.carrotsCollected < 0 then
				gameOver=true
				player.carrotsCollected = 0
			end
			
			if (rats.list[i].direction == "left") then
				rats.list[i].x=(0-ratWidth*1.5)
			else
				rats.list[i].x=(720+ratWidth*1.5)
			end
			rats.list[i].speed=math.random(ratSpeed-25,ratSpeed)
			rats.list[i].y=(math.random(0,570-ratHeight))
			pause=false
		elseif (rats.list[i].x>750) and (rats.list[i].direction=="left") and pause == false then
			pause=true
			
			if (rats.list[i].direction == "left") then
				rats.list[i].x=(0-ratWidth*1.5)
			else
				rats.list[i].x=(720+ratWidth*1.5)
			end
			rats.list[i].speed=math.random(ratSpeed-25,ratSpeed)
			rats.list[i].y=(math.random(0,570-ratHeight))
			pause=false
		elseif (rats.list[i].x<-30) and (rats.list[i].direction=="right") and pause == false then
			pause=true
			
			if (rats.list[i].direction == "left") then
				rats.list[i].x=(0-ratWidth*1.5)
			else
				rats.list[i].x=(720+ratWidth*1.5)
			end
			rats.list[i].speed=math.random(ratSpeed-25,ratSpeed)
			rats.list[i].y=(math.random(0,570-ratHeight))
			pause=false
		end
	end
end

function rats.draw()
	for i,v in pairs (rats.list) do
		if rats.list[i].direction == "left" then
			love.graphics.draw(ratSprite, rats.list[i].x, rats.list[i].y)
		elseif rats.list[i].direction == "right" then
			love.graphics.draw(rightRatSprite, rats.list[i].x, rats.list[i].y)
		end
	end
end

--[[-----------------------------------------------------
---------------------------------------------------------
---------------DIFFICULTY BASED FUNCTIONS----------------
---------------------------------------------------------
--]]-----------------------------------------------------

function ratCheck(howMany)
	maxRats=(math.ceil(howMany)/100)
	if maxRats==5 and fourHundredReached==false then
		fourHundredReached=true
		rats.list[2].direction="right"
	elseif maxRats==6 and fiveHundredReached==false then
		fiveHundredReached=true
		rats.list[4].direction="right"	
	end
	ratSpeed=200+maxRats*50
	if ratSpeed>300 then
		ratSpeed=300
	end
	if (#(rats.list) < maxRats) and (#(rats.list) < absMaxRats) then
		local inte=(#(rats.list)+1)
		rats.list[inte] = {}
		rats.list[inte].x = (-200-ratWidth*1.5)
		rats.list[inte].speed=math.random(ratSpeed-25,ratSpeed)
		rats.list[inte].y=(math.random(0,570-ratHeight))
		rats.list[inte].direction="left"
	end
	
end

function rhinoTime()
	if maxRats>=4 then
		return true
	else
		return false
	end
end

function oneOnionTime()
	if maxRats>=7 then
		return true
	else
		return false
	end
end

function twoOnionTime()
	if maxRats>=8 then
		return true
	else
		return false
	end
end

function moreRatsTime()

end

src/player.lua
player={}

function player.load()
	player.x = 200
	player.y = 200
    player.speed = PlayerSpeed
    player.name = "Cat"
    player.level = 1
	player.carrotsCollected = 0
	player.sprite = love.graphics.newImage("images/cat.png")
end

function player.update(dt)
	if love.keyboard.isDown('d') or love.keyboard.isDown('right') then -- movement
		if ((player.x) < (720-70)) then
			player.x = player.x + player.speed*dt
			player.sprite = love.graphics.newImage("images/catright.png")
		end
	end
	if love.keyboard.isDown('a') or love.keyboard.isDown('left') then -- movement
		if ((player.x) > (0)) then
			player.x = player.x - player.speed*dt
			player.sprite = love.graphics.newImage("images/catleft.png")
		end
	end
	if love.keyboard.isDown('w') or love.keyboard.isDown('up') then -- movement
		if ((player.y) > (0)) then
			player.y = player.y - player.speed*dt
			player.sprite = love.graphics.newImage("images/catup.png")
		end
	end
	if love.keyboard.isDown('s') or love.keyboard.isDown('down') then -- movement
		if ((player.y) < (570-76)) then
			player.y = player.y + player.speed*dt
			player.sprite = love.graphics.newImage("images/catdown.png")
		end
	end	
	if love.keyboard.isDown('escape') then
		backgroundMusic:stop()
		love.window.close()
		love.event.quit()
	end
	if love.keyboard.isDown("k") and (sounds.isPlaying == false) then -- sound
		sounds.isPlaying=true
		love.audio.play(sounds.meow)
		player.sprite = love.graphics.newImage("images/cat.png")
		sounds.isPlaying=false
	end
end

function player.draw()
	love.graphics.setNewFont(11)
	love.graphics.setColor(1,1,1)
	love.graphics.draw(player.sprite, player.x, player.y)
	love.graphics.setColor(0,0,0)
	love.graphics.print("Carrots:" .. player.carrotsCollected, player.x, player.y+80)
	love.graphics.setNewFont(24)
	love.graphics.setColor(1,1,1)
	love.graphics.setFont(ScoreFont)
	love.graphics.print("Total Carrots : " .. TOTALcarrotsCollected, 250, 10)
	if (OnionCounter>0) then
		love.graphics.print("Onions : " .. OnionCounter, 450, 10)
	end
end

src/rhinos.lua
rhinos={}
rhinos.list={}

rhinoUpSprite=love.graphics.newImage("images/rhinoup.png")
rhinoDownSprite=love.graphics.newImage("images/rhinodown.png")
rhinoWidth=rhinoUpSprite:getWidth()
rhinoHeight=rhinoUpSprite:getHeight()
--canPlayCollision=true

function rhinos.load()
	rhinos.list[1]={}
	rhinos.list[1].x=(math.random(0,720-rhinoWidth))
	rhinos.list[1].y=(570+math.random(300,400))
	rhinos.list[1].speed=50
	rhinos.list[1].direction="up"
end

function rhinos.update(dt)
	if (rhinos.list[1].direction=="up") then
		rhinos.list[1].y = (rhinos.list[1].y - rhinos.list[1].speed*dt)
	else
		rhinos.list[1].y = (rhinos.list[1].y + rhinos.list[1].speed*dt)
	end
	
	if (rhinos.list[1].y<(-401)) and (rhinos.list[1].direction=="up") then
		rhinos.list[1].x=(math.random(0,720-rhinoWidth))
		rhinos.list[1].y=(0-math.random(300,400))
		rhinos.list[1].direction = "down"
	elseif (rhinos.list[1].y>(1000)) and (rhinos.list[1].direction=="down") then
		rhinos.list[1].x=(math.random(0,720-rhinoWidth))
		rhinos.list[1].y=(570+math.random(300,400))
		rhinos.list[1].direction = "up"
	end
	
	if CheckCollision(rhinos.list[1].x,rhinos.list[1].y,rhinoWidth,rhinoHeight,player.x,player.y,player.sprite:getWidth(),player.sprite:getHeight()) then
		player.speed=0
		beingTrampled=5
		--if canPlayCollision==true then
			collisionSound:play()
			--canPlayCollision=false
		--end
	else
		--canPlayCollision=true
		beingTrampled=1
	end
	
end

function rhinos.draw()
	if rhinos.list[1].direction=="up" then
		love.graphics.draw(rhinoUpSprite, rhinos.list[1].x, rhinos.list[1].y)
	elseif rhinos.list[1].direction=="down" then
		love.graphics.draw(rhinoDownSprite, rhinos.list[1].x, rhinos.list[1].y)
	end
	--print("X :: " .. (rhinos.list[1].x) .. ": Y :: " .. (rhinos.list[1].y))
end

src/carrots.lua
carrots={}
carrots.list={}
maxCarrots=5
accelVarLow=200
accelVarHigh=350
pointsPerCarrot=1
goldenCarrotChance=21212
goldenGot=false
countingUp=0

function carrots.load()
	TOTALcarrotsCollected = 0
	
	carrotSprite=love.graphics.newImage("images/carrot.png")
	carrotWidth=carrotSprite:getWidth()
	carrotHeight=carrotSprite:getHeight()
	
	onionSprite=love.graphics.newImage("images/onion.png")
	onionWidth=onionSprite:getWidth()
	onionHeight=onionSprite:getHeight()
	
	for i=1,maxCarrots do
		carrots.list[i] = {}
		carrots.list[i].x=math.random(0,(720-carrotWidth))
		carrots.list[i].y=(0-carrotHeight*(math.random(10,15)/10))--math.random(0,(570-carrotHeight))
		carrots.list[i].accel=math.random(accelVarLow,accelVarHigh)
		carrots.list[i].golden=false
		carrots.list[i].isOnion=false
	end
	
end

function carrots.update(dt)
	for i,v in pairs (carrots.list) do
		carrots.list[i].y = (carrots.list[i].accel*dt + carrots.list[i].y)
	end

	for i,v in pairs (carrots.list) do
		if CheckCollision(carrots.list[i].x,carrots.list[i].y,carrotWidth,carrotHeight,player.x,player.y,player.sprite:getWidth(),player.sprite:getHeight()) and pause == false then
			pause=true
			
			if (carrots.list[i].isOnion==false) then
				player.carrotsCollected=player.carrotsCollected+pointsPerCarrot
				TOTALcarrotsCollected=TOTALcarrotsCollected+pointsPerCarrot
			else
				OnionCounter=OnionCounter+pointsPerCarrot
			end
			
			carrots.list[i].x=math.random(0,(720-carrotWidth))
			carrots.list[i].y=(0-carrotHeight*(math.random(10,15)/10))
			carrots.list[i].accel=math.random(accelVarLow,accelVarHigh)
			if (oneOnion==true) and (i==2) then --turns a carrots into onion  #1
				carrots.list[i].isOnion=true
			elseif (twoOnion==true) and (i==3) then --turns a carrots into onion #2
				carrots.list[i].isOnion=true
			end
			if carrots.list[i].golden == true then
				goldenGot=true
			end
			if math.random(0,goldenCarrotChance) == 42 and (carrots.list[i].isOnion == false) then
				carrots.list[i].golden=true
			else 
				carrots.list[i].golden=false
			end
			pause=false
		elseif (carrots.list[i].y>650) and pause == false then
			pause=true
			
			carrots.list[i].x=math.random(0,(720-carrotWidth))
			carrots.list[i].y=(0-carrotHeight*(math.random(10,15)/10))
			carrots.list[i].accel=math.random(accelVarLow,accelVarHigh)
			carrots.list[i].golden = false
			if (oneOnion==true) and (i==2) then --turns a carrot into onion  #1
				carrots.list[i].isOnion=true
			elseif (twoOnion==true) and (i==3) then --turns a carrot into onion #2
				carrots.list[i].isOnion=true
			end
			if math.random(0,goldenCarrotChance) == 42 and (carrots.list[i].isOnion == false)then
				carrots.list[i].golden=true
			end
			pause=false
		end
	end
end

function carrots.draw()
	for i,v in pairs (carrots.list) do
		if (carrots.list[i].golden == false) and (carrots.list[i].isOnion == false) then
			love.graphics.draw(carrotSprite, carrots.list[i].x, carrots.list[i].y)
		elseif (carrots.list[i].golden == true) and (carrots.list[i].isOnion == false) then
			love.graphics.draw(goldenCarrotSprite, carrots.list[i].x, carrots.list[i].y)
		else
			love.graphics.draw(onionSprite, carrots.list[i].x, carrots.list[i].y)
		end
	end
end

src/main.lua
require "player"
require "carrots"
require "rats"
require "rhinos"
require "bats"
socket = require("socket")

function sleep(sec)
    socket.sleep(sec)
end
DefaultFont=love.graphics.newFont(14)
ScoreFont = love.graphics.newImageFont("images/imagefont.png",
    " abcdefghijklmnopqrstuvwxyz" ..
    "ABCDEFGHIJKLMNOPQRSTUVWXYZ0" ..
    "123456789.,!?-+/():;%&`'*#=[]\"")

sounds = {}
pause=false
gameOver=false
PlayerSpeed=365
gameOverScreen=love.graphics.newImage("images/gameover.png")
easterEggScreen=love.graphics.newImage("images/garbage.png")
goldenCarrotSprite=love.graphics.newImage("images/goldenC.png")
backgroundImage=love.graphics.newImage("images/background2.png")
beingTrampled=1
oneOnion=false 
twoOnion=false
OnionCounter=0
BatsReleased=false
BatsProb=15

function love.load() -- This code gets executed once
	math.randomseed( os.time() )

    love.graphics.setBackgroundColor(199/255,189/255,100/255) -- Background
	
	player.load()
	
	bats.load()
	
    sounds.meow = love.audio.newSource("audio/meow.mp3","stream")
	backgroundMusic = love.audio.newSource("audio/background.ogg","static")
	collisionSound = love.audio.newSource("audio/collision.ogg","static")
	backgroundMusic:setLooping(true)
	backgroundMusic:play()
    sounds.isPlaying = false
	
	carrots.load()
	
	rats.load()
	
	rhinos.load()
end

function love.update(dt)-- this code gets executed every "tick"
	if dt>1/30 then
		dt=1/30
	end
	player.update(dt)
	if gameOver==false then
		
		player.speed=PlayerSpeed
		if (rhinoTime() == true) then
			rhinos.update(dt)
		end
		
		if (oneOnionTime() == true) then
			oneOnion=true
		end
		
		if (twoOnionTime() == true) then
			twoOnion=true
		end
		
		if OnionCounter>1 and BatsReleased==false then
			local thingy = math.ceil(OnionCounter/7)
			if thingy > 7 then 
				thingy = 7 
			end
			countTo=(8-thingy)
			countingUp=countingUp+dt
			if countingUp>=countTo then
				BatsReleased=true
			end
			
		end
		
		player.update(dt)
		
		ratCheck(TOTALcarrotsCollected)
		carrots.update(dt)
		
		rats.update(dt)
		
		if BatsReleased == true then
			bats.update(dt)
		end
		
	else
		for i,v in pairs (carrots.list) do
			table.remove(carrots.list,i)
		end
	end
end

-- Collision detection function;
-- Returns true if two boxes overlap, false if they don't;
-- x1,y1 are the top-left coords of the first box, while w1,h1 are its width and height;
-- x2,y2,w2 & h2 are the same, but for the second box.
function CheckCollision(x1,y1,w1,h1, x2,y2,w2,h2)
  return x1 < x2+w2 and
         x2 < x1+w1 and
         y1 < y2+h2 and
         y2 < y1+h1
end

function CheckCollisionSides(x1,y1,w1,h1, x2,y2,w2,h2)
  return x1 < x2+w2 and
         x2 < x1+w1
end

function CheckCollisionTopBot(x1,y1,w1,h1, x2,y2,w2,h2)
  return y1 < y2+h2 and
         y2 < y1+h1
end

function love.draw() -- this code gets executed after every update. The only difference between this and the update is that this can draw.
	love.graphics.setNewFont(10)
	love.graphics.setColor(1,1,1)
	love.graphics.draw(backgroundImage,0,0)
	love.graphics.setColor(0,0,0)
	love.graphics.print("FPS: " .. love.timer.getFPS(), 10, 10)
	if gameOver == true then
		love.graphics.setColor(1,1,1)
		love.graphics.draw(gameOverScreen,0,0)
		love.graphics.setColor(1,1,1)
		love.graphics.setFont(ScoreFont)
		love.graphics.print("Total Score : " .. TOTALcarrotsCollected + math.floor(OnionCounter/7), 280, 10)
		
	elseif goldenGot==true then
		love.graphics.setColor(1,1,1)
		love.graphics.draw(easterEggScreen,0,0)
		backgroundMusic:stop()
		--sleep(0.2)
		--love.window.close()
		--love.event.quit()
	elseif gameOver == false then
		
		player.draw()
		
		love.graphics.setColor(1,1,1)
		
		rats.draw()
		
		if (rhinoTime() == true) then
			rhinos.draw(dt)
		end
		
		carrots.draw()
		
		if BatsReleased==true then
			bats.draw()
		end
	end
end


function love.mousepressed(x,y,button)

end

function love.mousereleased(x,y,button)

end

function love.keypressed( key )

end

